### PR TITLE
Configurable order state machine  and reduce the number of possible transitions

### PR DESCRIPTION
**Pull request by mamhoff**
This PR can be divided into two logical parts that are interdependent and
breaking each one in their way. These parts are described in detail later.

**Make Order state machine configurable**

In 7c4f82c21, all relevant state machines except for the order one have
been extracted into modules. This performs the same exercise for the
order state machine. A lot of this code is deeply interlinked with the
rest of the system, so you'd have to take lots of care when replacing
this; but it should be possible nonetheless.

The new state machine is named `Spree::Core::StateMachines::Order` and
includes the same code of the old one.

The newly named state machine is active on new stores generated from
scratch and in Solidus tests, but old stores will still use the old one,
unless the preference `Spree::Config.use_legacy_order_state_machine` is
explicitly set to `false`.

When using the old named state machine a deprecation warning will be
printed at boot time, as Solidus 3.0 will start using the new order state
machine by default.

**Reduce the number of new order state machine transitions** 

This removes many, many transitions from the Order state machine definitions
included in `Spree::Core::StateMachines::Order`, allowing us to reason more
clearly about what the system is supposed to be doing and when.

This new state machine definition is not active yet on existing stores, but
will become standard from Solidus 3.0.

The  legacy state machine definitions from `Spree::Order::Checkout` are left
untouched, so old stores can continue to use them.

A bunch of transitions have been removed, and now an order can transition to:

  * Canceled only when Complete;
  * Awaiting Return only when Complete;
  * Returned only when Complete or Awaiting Return.

This change was initially proposed with https://github.com/solidusio/solidus/pull/3532 but for convenience it 
has been included directly here as the original PR,  being a breaking change, was 
not mergeable in isolation without the premise of the first part of this PR. Please
consider taking a look at the original PR for the interesting discussion.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [ ] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: core/lib/spree/core/state_machines.rb [hunk 1]
```diff
@@ -65,6 +66,15 @@ def shipment
         @shipment.constantize
       end
 
+      def order
+        @shipment ||= begin
```
Comment (kennyadsl) [inline core/lib/spree/core/state_machines.rb]: should be `@order` here

File: core/lib/spree/core/state_machines/order.rb [hunk 2]
```diff
@@ -0,0 +1,245 @@
+# frozen_string_literal: true
+
+module Spree
+  module Core
+    class StateMachines
+      module Order
```
Comment (kennyadsl) [inline core/lib/spree/core/state_machines/order.rb]: I don't think we can change the name of the module without a deprecation proxy. Anyway, I'm also ok leaving `Checkout` as its name for now. Is there a specific reason to change it within this PR?

**Comment by aldesantis** [comment]
@mamhoff it would be awesome to get this in 2.11, which is the last release of the 2.x branch. The idea is to merge this in 2.11 and deprecate the old state machine module, then remove it completely in 3.0.

**Comment by mamhoff** [comment]
Sure, go ahead! I'm really happy this is moving forward.

Andrea Longhi <notifications@github.com> schrieb am Fr., 29. Mai 2020,
16:40:

**Comment by seand7565** [comment]
Quick note here, to reflect discussion on the issue: When this is merged, we can close #1576

File: core/lib/spree/core/state_machines/order.rb [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        def passed_checkout_step?(step)
+          has_checkout_step?(step) && checkout_step_index(step) < checkout_step_index(state)
+        end
+
+        def checkout_step_index(step)
+          checkout_steps.index(step).to_i
+        end
+
+        def can_go_to_state?(state)
+          return false unless has_checkout_step?(self.state) && has_checkout_step?(state)
```
[the remark this answers is not in the document]
Comment (spaghetticode) [reply core/lib/spree/core/state_machines/order.rb:241]: fixed

File: core/lib/spree/core/state_machines/order.rb [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            end
+          end
+
+          def remove_transition(options = {})
+            removed_transitions << options
+            next_event_transitions.delete(find_transition(options))
+          end
+
+          def find_transition(options = {})
+            return nil if options.nil? || !options.include?(:from) || !options.include?(:to)
```
[the remark this answers is not in the document]
Comment (spaghetticode) [reply core/lib/spree/core/state_machines/order.rb:189]: fixed

File: core/lib/spree/core/state_machines/order.rb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+          end
+
+          def removed_transitions
+            @removed_transitions ||= []
+          end
+        end
+
+        def checkout_steps
+          steps = self.class.checkout_steps.each_with_object([]) { |(step, options), checkout_steps|
+            next if options.include?(:if) && !options[:if].call(self)
```
[the remark this answers is not in the document]
Comment (spaghetticode) [reply core/lib/spree/core/state_machines/order.rb:219]: fixed

File: core/lib/spree/core/state_machines/order.rb (hunk 2 above)
Comment (spaghetticode) [reply core/lib/spree/core/state_machines/order.rb:6]: I'm resolving this conversation as I think it's not relevant anymore now that we decided to keep the old state machine around for a while.

@kennyadsl feel free to reopen if it makes sense to you.

**Comment by kennyadsl** [comment]
I hope to be able to talk about this in our Core Team meeting this week. 
Anyway, we just merged https://github.com/solidusio/solidus/pull/3658 and this probably requires some extra work to also backport the changes introduced there.

**Comment by spaghetticode** [comment]
I explored the idea of adding a deprecation to the constant `Spree::Core::StateMachines::Order` but both `ActiveSupport::Deprecation::DeprecatedConstantAccessor` and `ActiveSupport::Deprecation::DeprecatedConstantProxy` require a replacement constant that will be used instead of the deprecated one, as rightfully predicted by @kennyadsl, so they don't fit here. In the end, I think we can be happy with the deprecation at https://github.com/solidusio/solidus/pull/3542/files#diff-b80f3499d79adcd431c150b9964453feR72.

**Pull request merged**